### ISSUE TITLE
Overdue is incorrect

### ISSUE THREAD

**Issue by Virsacer** [Icinga/icingadb#546]

**Comment by nilmerg** [comment]
Icinga DB is calculating this, hence I transferred this here.

**Comment by julianbrost** [comment]
Definitely looks like an oversight. Might very well be the case that this requires changes in icinga2 due to incorrect or incomplete information in Redis.

@Al2Klimov Can you please have a look at this?

**Comment by Al2Klimov** [comment]
Luckily at least "even" the monitoring module doesn’t handle it properly either.

**Comment by Al2Klimov** [comment]
@julianbrost Shall I extend Checkable#GetNextUpdate() – which is the source of truth – so that it takes the check_period into account (if possible)?

**Comment by Al2Klimov** [comment]
@nilmerg It is with great pleasure to inform you:

> A `check_period` is defied for the service but it is (...) not even displayed on the service page.

This is web's construction area.

**Comment by nilmerg** [comment]
> A check_period is defied for the service but it is (...) not even displayed on the service page.

@Virsacer Please create a separate issue.

**Comment by julianbrost** [comment]
> @julianbrost Shall I extend Checkable#GetNextUpdate() – which is the source of truth – so that it takes the check_period into account (if possible)?

Doesn't this basically happen due to some discrepancy between the check scheduling logic and that function? Please check if it's possibly to just share the same code for both rather than implementing the `check_period` check in a second place.

Apart from that: you also have to make sure that the value in Redis is actually updated, I'm not sure if enter/leave check_period would already trigger an update there. And please also look if something similar happens for other situations where Icinga won't perform checks (manually disable active checks, active checks disabled by a dependency).

**Comment by Al2Klimov** [comment]
> Doesn't this basically happen due to some discrepancy between the check scheduling logic and that function? Please check if it's possibly to just share the same code for both rather than implementing the `check_period` check in a second place.

No, I don’t think so. The checker basically says _if now is not inside check_period, next_check += check_interval_. This even makes sense to let timeperiod updates take effect ASAP.

> 
> Apart from that: you also have to make sure that the value in Redis is actually updated, I'm not sure if enter/leave check_period would already trigger an update there. And please also look if something similar happens for other situations where Icinga won't perform checks (manually disable active checks, active checks disabled by a dependency).

1. I consider the update not necessary. If the check should have been at 14:00, but was missing and the TP ended at 15:00, the check is still overdue IMAO.
2. Regarding passive-only checks: I see no significant difference for our case. Dependencies are a problem due to no ETA, but IMAO web could show UNREACHABLEs as not-overdue unconditionally. (@nilmerg Correct me if anything.)

**Comment by Al2Klimov** [comment]
## My suggested change-set

```patch
--- lib/icinga/checkable.cpp
+++ lib/icinga/checkable.cpp
@@ -227,7 +227,18 @@ Timestamp Checkable::GetNextUpdate() const
                latency = 0.0;
        }

-       return (GetEnableActiveChecks() ? GetNextCheck() : (cr ? cr->GetExecutionEnd() : Application::GetStartTime()) + interval) + interval + 2 * latency;
+       auto nextUpdate ((GetEnableActiveChecks() ? GetNextCheck() : (cr ? cr->GetExecutionEnd() : Application::GetStartTime()) + interval) + interval + 2 * latency);
+       auto checkPeriod (GetCheckPeriod());
+
+       if (checkPeriod && !checkPeriod->IsInside(nextUpdate)) {
+               auto nextTransition (checkPeriod->FindNextTransition(nextUpdate));
+
+               if (nextTransition > nextUpdate) {
+                       nextUpdate = nextTransition;
+               }
+       }
+
+       return nextUpdate;
 }

 void Checkable::NotifyFixedDowntimeStart(const Downtime::Ptr& downtime)
```

**Comment by julianbrost** [comment]
> The checker basically says _if now is not inside check_period, next_check += check_interval_.

But shouldn't this just result in next update being updated in Redis as well? The value is based on the next check time after all.

Just tried that: if you manually reschedule the check into the future, it goes overdue as well.

> 1. I consider the update not necessary.

I've interpreted "so that it takes the check_period into account" as a simple check if it's inside at the moment. Then something like this would have been necessary.

> 2. [...] but IMAO web could show UNREACHABLEs as not-overdue unconditionally

Sure it could, but if something isn't checked intentionally, it shouldn't be marked as overdue anywhere, rather than just touching up things in the frontend.
